tests/cpp/program/graph_test.cpp

```diff
@@ -0,0 +1,52 @@
+#include "gtest/gtest.h"
+#include "taichi/ir/ir_builder.h"
+#include "taichi/ir/statements.h"
+#include "taichi/inc/constants.h"
+#include "taichi/program/program.h"
+#include "tests/cpp/program/test_program.h"
+#include "taichi/program/graph.h"
+#include "tests/cpp/ir/ndarray_kernel.h"
+#ifdef TI_WITH_VULKAN
+#include "taichi/backends/vulkan/vulkan_loader.h"
+#endif
+
+using namespace taichi;
+using namespace lang;
+#ifdef TI_WITH_VULKAN
+TEST(GraphTest, SimpleGraphRun) {
+  // Otherwise will segfault on macOS VM,
+  // where Vulkan is installed but no devices are present
+  if (!vulkan::is_vulkan_api_available()) {
+    return;
+  }
+  TestProgram test_prog;
+  test_prog.setup(Arch::vulkan);
+
+  const int size = 10;
+
+  auto ker1 = setup_kernel1(test_prog.prog());
+  auto ker2 = setup_kernel2(test_prog.prog());
+
+  auto g = std::make_unique<Graph>("test");
+  auto seq = g->seq();
+  auto arr_arg = aot::Arg{
+      "arr", PrimitiveType::i32.to_string(), aot::ArgKind::NDARRAY, {}};
+  seq->dispatch(ker1.get(), {arr_arg});
+  seq->dispatch(ker2.get(),
+                {arr_arg, aot::Arg{"x", PrimitiveType::i32.to_string(),
+                                   aot::ArgKind::SCALAR}});
+  g->compile();
```

Comment:
Please remember to separate out the builder :-) https://refactoring.guru/design-patterns/builder